presto-common/src/main/java/com/facebook/presto/common/block/AbstractArrayBlock.java

```diff
@@ -0,0 +1,318 @@
+            return createArrayBlockInternal(0, length, newValueIsNull, newOffsets, newValues);
+        }
+
+        // Second traversal will use the newOffsets and total element positions to
+        // populate a correctly sized values position selection array
+        int[] elementPositions = new int[totalElements];
+        int currentOffset = 0;
+        for (int i = 0; i < length; i++) {
+            int elementsLength = newOffsets[i + 1] - newOffsets[i];
+            if (elementsLength > 0) {
+                int valuesStartOffset = getOffset(positions[i + offset]);
+                for (int j = 0; j < elementsLength; j++) {
+                    elementPositions[currentOffset++] = valuesStartOffset + j;
+                }
+            }
+        }
+        verify(currentOffset == elementPositions.length, "unexpected element positions");
+        Block newValues = getRawElementBlock().copyPositions(elementPositions, 0, elementPositions.length);
+        return createArrayBlockInternal(0, length, newValueIsNull, newOffsets, newValues);
+    }
+
+    @Override
+    public Block getRegion(int position, int length)
+    {
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, position, length);
+
+        return createArrayBlockInternal(
+                position + getOffsetBase(),
+                length,
+                getValueIsNull(),
+                getOffsets(),
+                getRawElementBlock());
+    }
+
+    @Override
+    public long getRegionSizeInBytes(int position, int length)
+    {
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, position, length);
+
+        int valueStart = getOffsets()[getOffsetBase() + position];
+        int valueEnd = getOffsets()[getOffsetBase() + position + length];
+
+        return getRawElementBlock().getRegionSizeInBytes(valueStart, valueEnd - valueStart) + ((Integer.BYTES + Byte.BYTES) * (long) length);
+    }
+
+    @Override
+    public long getRegionLogicalSizeInBytes(int position, int length)
+    {
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, position, length);
+
+        int valueStart = getOffsets()[getOffsetBase() + position];
+        int valueEnd = getOffsets()[getOffsetBase() + position + length];
+
+        return getRawElementBlock().getRegionLogicalSizeInBytes(valueStart, valueEnd - valueStart) + ((Integer.BYTES + Byte.BYTES) * (long) length);
+    }
+
+    @Override
+    public long getApproximateRegionLogicalSizeInBytes(int position, int length)
+    {
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, position, length);
+
+        int valueStart = getOffset(position);
+        int valueEnd = getOffset(position + length);
+
+        return getRawElementBlock().getApproximateRegionLogicalSizeInBytes(valueStart, valueEnd - valueStart) + (Integer.BYTES + Byte.BYTES) * length;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).